bin/warnhist

```diff
@@ -67,29 +86,196 @@ sub do_help {
     print "\t\t<name list> is a comma separated list, with no spaces\n";
     print "\t\tFor example: 'H5LTparse' or 'H5LTparse,H5LTanalyze'\n";
+    print "\t-d\tCompute warning density for compiled source files.  Paths to the\n";
+    print "\t\troot of a directory containing source may be provided with the\n";
+    print "\t\t'-p <path list>' option.  If the path list is given, only those\n";
+    print "\t\tdirectories are scanned for source files.  If the path list\n";
+    print "\t\toption is not given, the current working directory is scanned.\n";
+    print "\t-p <path list>\tPaths to search for compiled files.  Compiled files\n";
+    print "\t\tare only used when computing warning density and are not\n";
+    print "\t\tnecessary for just analyzing warnings in build output.\n";
+    print "\t\t<path list> is a comma separated list, with no spaces\n";
+    print "\t\tFor example: '/home/koziol/hdf5' or '.,~/dev/hdf5,~/dev/build'\n";
+    print "\t-q\tSuppress warning output\n";
     print "\tfile\tFilename containing build output\n";
     print "\t\tIf no file is given, standard input is used.\n";
     exit;
 }
 
+# Count # of lines in a file
+sub line_count {
+    my ($name) = @_;
+#print "name = '$name'\n";
+    my $tmp;
+    my $lines = 0;
+
+    open (FILE, $name) or die "Can't open '$name': $!";
+    $lines++ while ($tmp = <FILE>);
+    close FILE;
+#print "$lines\n";
+
+    return $lines;
+}
+
+# Recursively search a directory hierarchy for source files
+# Adds results to the global %c_files, %cpp_files, and %fort_files hashes
+sub parse_tree {
+    my ($root_path) = @_;
+#print "root_path = $root_path\n";
+
+    my $path_checker = sub {
+        my $name = $File::Find::name;
+        if (-f $name) {
+	    my $bn = basename($name);
+
+	    # Check for different kinds of source files
+	    # Use lists here: https://gist.github.com/ppisarczyk/43962d06686722d26d176fad46879d41#file-programming_languages_extensions-json
+
+            # FORTRAN source file
+	    if($bn =~ /.*(\.f90)|(\.f)|(\.f03)|(\.f08)|(\.f77)|(\.f95)|(\.for)|(\.fpp)$/i) {
+		$bn =~ s/(\.f90)|(\.f)|(\.f03)|(\.f08)|(\.f77)|(\.f95)|(\.for)|(\.fpp)$//ig;
+		if(!exists $fort_files{$bn}) {
+		    $fort_files{$bn} = [ $name ];
+		} else {
+		    push @{ $fort_files{$bn} }, $name;
+		}
+            # C++ source file
+	    } elsif($bn =~ /.*(\.cpp)|(\.c\+\+)|(\.cc)|(\.cp)|(\.cxx)$/i) {
+		$bn =~ s/(\.cpp)|(\.c\+\+)|(\.cc)|(\.cp)|(\.cxx)$//ig;
+		if(!exists $cpp_files{$bn}) {
+		    $cpp_files{$bn} = [ $name ];
+		} else {
+		    push @{ $cpp_files{$bn} }, $name;
+		}
+            # C source file
+	    } elsif($bn =~ /.*(\.c)$/i) {
+		$bn =~ s/(\.c)$//g;
+		if(!exists $c_files{$bn}) {
+		    $c_files{$bn} = [ $name ];
+		} else {
+		    push @{ $c_files{$bn} }, $name;
+		}
+	    }
+      }
+    };
+    find($path_checker, $root_path);
+
+#print Dumper \%c_files;
+#print Dumper \%cpp_files;
+#print Dumper \%fort_files;
+}
+
+sub count_file_loc {
```

Comment:
Locate source files, counting their lines of code.  Handles multiple source files correctly, as long as the number of compiles equals the number of times a file of that name is found in the directory hierachy(s).